awscli/customizations/history/list.py

```diff
@@ -0,0 +1,499 @@
+           replaced with ...
+        3) If a parameter is about to be added to the current line (--foo) we
+           assume that the next argument is its value (--region us-west-2)
+           and we try to fit them on the same line together. If they will not
+           fit together at the end of the line, and we have already added
+           content to this line we will omit the parameter from this line in
+           the hope that they will fit together on the next line.
+        4) If we are adding a parameter and there is no content already on the
+           line and it still does not fit with its value then it never will
+           since this is a fresh line. We put the paramater on the current line
+           and if it looks like the next argument is a value it will be
+           indented an extra two spaces to show visually that it belongs to
+           the paramater on the line above it.
+        """
+        space_left = self._space
+        line = ""
+        while self.has_more_content() and space_left > 0:
+            candidate = self._value.pop(0)
+            param = self._is_param(candidate)
+            candidate = self._indent_candidate(line, candidate, param)
+            fits = self._does_fit_on_line(line, candidate, space_left)
+
+            if self._end_line_early(line, candidate, param, fits, space_left):
+                break
+
+            if fits:
+                line, space_left = self._add_to_line(
+                    line, candidate, param, space_left)
+            else:
+                if not line:
+                    # Nothing got added to the line yet which means this
+                    # candidate is too long to fit on a single line and needs
+                    # to be truncated and made to fit.
+                    line, space_left = self._truncate_and_add_to_line(
+                        line, candidate, param, space_left)
+                else:
+                    # Naturally ran out of space and cannot fit the next
+                    # candidate into this row. We will put the candidate back.
+                    # Since our loop at the top checks for space and an element
+                    # remaining it will loop again if we put the candidate back
+                    # so we will also break here to prevent an infinite loop.
+                    self._value.insert(0, candidate)
+                    break
+        line = self._indent_line_if_needed(line)
+        return line
+
+
+class TableCellFormatterFactory(object):
+    _COL_TO_FORMATTER_TYPE = {
+        'Command Id': StringTableCellFormatter,
+        'Time': StringTableCellFormatter,
+        'Arguments': ArgumentTableCellFormatter,
+        'Return Code': StringTableCellFormatter,
+    }
+
+    def create(self, col, value, space):
+        cls = self._COL_TO_FORMATTER_TYPE[col]
+        formatter = cls(value, space)
+        return formatter
+
+
+class TableStyler(object):
```

Comment:
There's already a table formatter used for the `--output table` format.  Why isn't that class being used?  Doesn't seem like we should be maintaining two separate classes.

Reply:
It doesn't really do what we want. I tried using that for awhile and I was just fighting the design since it does all sorts of magic under the hood for arbitrary payload shapes whereas we have a fixed schema and a fixed order we want to show, with specific colors. I found it much easier to get what I wanted when I removed the existing one and started from scratch.

Reply:
That doesn't make sense to me.  It doesn't do anything payload shapes.  It doesn't know anything about shapes or models.  Are you sure you're looking at `TableFormatter`?

I suppose I was being too subtle with my previous feedback.  Please consolidate this.  Unless you have concrete issues with why this can't be done (or point me at a branch where you tried), we shouldn't be maintaining two different ways to draw tables.

Either that or drop the table format.  I personally wish we would have have something similar to the the text based output so I can manipulate it with `cut`, `awk`, etc.  That's how I was hoping to use the list command.